add initial codeberg support

laumann opened the pull request:
This allows
```
pram https://codeberg.org/gentoo/gentoo/pulls/1
```
to work as expected, at least for a PR with a single patch (haven't tested with multiple patches yet).

The [patch](https://codeberg.org/gentoo/gentoo/pulls/1.patch)

```patch
From 096358ce85bd9c77ba5fb8eec6f9891683930295 Mon Sep 17 00:00:00 2001
From: Thomas Bracht Laumann Jespersen <t@laumann.xyz>
Date: Tue, 24 Jun 2025 15:18:47 +0200
Subject: [PATCH] dev-libs/tlsh: raise minimum required cmake version

Closes: https://bugs.gentoo.org/953851
Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>
---
 dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch b/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch
index d30363de6d74..601f3d899b72 100644
--- a/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch
+++ b/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch
@@ -5,7 +5,7 @@
  # OF THE POSSIBILITY OF SUCH DAMAGE.
  
 -cmake_minimum_required(VERSION 2.6)
-+cmake_minimum_required(VERSION 2.8.5)
++cmake_minimum_required(VERSION 3.10)
  
  project(TLSH)
  
-- 
2.47.3
```
is applied like so, with added `Closes` and `Signed-off-by` trailers
```diff
commit d8396bd6af89a12257abff33c733aeac3389b911
Author: Thomas Bracht Laumann Jespersen <t@laumann.xyz>
Date:   Tue Jun 24 15:18:47 2025 +0200

    dev-libs/tlsh: raise minimum required cmake version
    
    Closes: https://bugs.gentoo.org/953851
    Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>
    Part-of: https://codeberg.org/gentoo/gentoo/pulls/1
    Closes: https://codeberg.org/gentoo/gentoo/pulls/1
    Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>

diff --git a/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch b/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch
index d30363de6..601f3d899 100644
--- a/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch
+++ b/dev-libs/tlsh/files/tlsh-4.8.2-gnuinstalldirs.patch
@@ -5,7 +5,7 @@
  # OF THE POSSIBILITY OF SUCH DAMAGE.
  
 -cmake_minimum_required(VERSION 2.6)
-+cmake_minimum_required(VERSION 2.8.5)
++cmake_minimum_required(VERSION 3.10)
  
  project(TLSH)
  ```

laumann committed:
add initial codeberg support

laumann commented:
The amount of special-casing for github in the script makes me wonder if this proposal is insufficient. I should certainly test a PR with multiple commits, but I'm not sure what else I need to consider here.

thesamesam commented:
There's https://codeberg.org/forgejo/forgejo/issues/7972 but that implies it _does_ support `Closes` already.

laumann commented:
> There's https://codeberg.org/forgejo/forgejo/issues/7972 but that implies it _does_ support `Closes` already.

Maybe the `Closes: ###` _has_ to be a PR number and a full URL is not supported?

eli-schwartz commented:
https://codeberg.org/forgejo/forgejo/issues/7972#issuecomment-7804976

parona-source committed:
Add support for Codebergs' api with tokens

No one has complained about rate limits yet, but its no doubt going to
be an issue in our post LLM era.

Signed-off-by: Alfred Wingate <parona@protonmail.com>

parona-source committed:
Support the same url types in Codeberg as GitHub already does

Only thing missing is getting compare patches from the api.

Signed-off-by: Alfred Wingate <parona@protonmail.com>

parona-source commented:
https://github.com/gentoo/pram/compare/master...parona-source:pram:codeberg

Hooked up the api as well. Only thing really lacking from Codeberg's api is getting the patches for compares.

laumann commented on pram.1:
```diff
@@ -75,6 +76,13 @@ string options:
+PRam will use the Codeberg REST API if a token is present at \fB~/.codeberg-token\fR.
+
+.UR https://docs.codeberg.org/advanced/access-token/
+See Codeberg documentation on how to create one
```
Should we include a comment about the required token permissions? Like, "the token must include the read:repository scope".

EDIT: reworded

parona-source replied:
That would be reasonable.

laumann committed:
Support appending Merges: trailer for Codeberg PRs

Make the default behaviour for merging Codeberg PRs to add "Merges:"
trailer instead of "Closes:" with the --no-merge option to force the
trailer to be "Closes:".

Signed-off-by: Thomas Bracht Laumann Jespersen <t@laumann.xyz>
